### PR TITLE
tracing: Add user-definable tracing interface

### PR DESCRIPTION
**Pull request by nslowell**
Create a new tracing option TRACING_USER that allows
the user to define certain user_sys_trace_... functions
to perform whatever work desired for tracing when
tasks are swiched in/out, during isr enter/exit, and when
cpu is idle.

This infrastructure can be useful for plugging into
locally defined tracing tools or any user-specific
debugging environment.

Signed-off-by: Nicholas Lowell <nlowell@lexmark.com>

### CODE REVIEW

File: subsys/tracing/include/tracing_user.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#define _TRACE_USER_H
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <init.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void user_sys_trace_thread_switched_in(struct k_thread *thread);
```
Comment (jukkar) [inline subsys/tracing/include/tracing_user.h:17]: The namespace of these functions look a bit weird, should be place the `user` at the end of the func name, so
`sys_trace_thread_switched_in_user()`?

Comment (nslowell) [reply subsys/tracing/include/tracing_user.h:17]: it follows the same naming convention as ctf, so there was precedent for this.

File: doc/guides/tracing/index.rst [hunk 2]
```diff
@@ -140,6 +140,24 @@ and exits (only distinguishes between idle thread, non idle thread and scheduler
 Enable this format with the :option:`CONFIG_TRACING_CPU_STATS` option.
 
 
+User-Defined Tracing
+====================
+
+This tracing format allows the user to define functions to perform any work desired
+when a task is switched in or out, when an interrupt is entered or exited, and when the cpu
+is idle.
+
```
Comment (nashif) [inline doc/guides/tracing/index.rst:149]: Can you please elaborate a bit on the usecase here and why you want to do that? Based on the above this seems like misuse of tracing to do something else not related to tracing, but maybe I am misunderstanding this.

Comment (nslowell) [reply doc/guides/tracing/index.rst:149]: Let me know if the below is sufficient.  If so, I can add it:

Examples include
 - simple toggling of GPIO for external scope tracing while minimizing extra cpu load
 - generating/outputting trace data in a non-standard or proprietary format that can not be supported by the other tracing systems

**Comment by nslowell** [comment]
I believe the failed test is unrelated and I never had a response from my last comment.

**Comment by nslowell** [comment]
@nashif please confirm changes

**Review by wentongwu — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
LGTM

**Review by nashif — DISMISSED** [review]
Can you provide a sample/test using this code please?

File: subsys/tracing/include/tracing_user.h (hunk 1 above)
Comment (nashif) [reply subsys/tracing/include/tracing_user.h:17]: I agree that user_sys_ is weird, where is this the case for ctf?

Comment (nslowell) [reply subsys/tracing/include/tracing_user.h:17]: @nashif 
in ctf_top.c: sys_trace_thread_switched_in() calls ctf_top_thread_switched_in() (where the prefix identifies the subsystem)

I was trying to copy that by having sys_trace_thread_switched_in() call user_sys_trace_thread_switched_in().

Would user_thread_switched_in() be an acceptble name to maintain the prefix precedence?

**Comment by bbolen** [comment]
> Can you provide a sample/test using this code please?

added a sample that uses the code which will work in qemu.

File: subsys/tracing/include/tracing_user.h (hunk 1 above)
Comment (bbolen) [reply subsys/tracing/include/tracing_user.h]: Moved the "user" to the end of the function names.  That seems to be a little more pleasing to the eyes.

**Comment by bbolen** [comment]
@nashif ping.  All issues should have been addressed.

**Pull request merged**